### ISSUE TITLE
Logging leaks private user data when a request fails

### ISSUE THREAD

**Issue by ajevans85** [http4s/http4s#1312]
If a request fails the full request object is written to logs by calling `toString` on the request object. This leads to leaking data such as Authentication headers with sensitive credentials being written to logs. 

At the moment there is no easy way to hook in to `Http1ServerStage.runRequest` which calls `internalServerError` resulting in the log message.

It's useful to dump the request for debugging purposes but this also imposes potential security issues. It would be great if some plugable error handler could be used so a user can choose to log or not and the contents of the log message.

Based on this any logging elsewhere in HTTP4s which calls Request.toString also should be refactored to allow overriding the error handling to prevent data leaks in to logging systems

**Comment by rossabaker** [comment]
It's common to wrap an `HttpService` with something that does `handleWith` (or use `MonadError` instance) to customize the handling of errors.

I strongly agree that we shouldn't be blindly logging headers.  (@ChristopherDavenport: the new logging middlewares should have an option to blacklist headers.)  Showing all the headers is a arguably a dangerous default for `request.toString`.

**Comment by ChristopherDavenport** [comment]
So to mirror gitter. The obvious solutions are to add filters such as 

```scala
type HeaderFilter = Headers => Headers
type EntityBodyFilter = EntityBody => EntityBody
def apply(logHeaders: Boolean, headerFilters: List[HeaderFilter], logBody: Boolean, entityBodyFilters: List[EntityBodyFilters])
```

Which opens us up to creating many filters for people to be able to apply easily. Both could use identity as the mempty for folding.

**Comment by ajevans85** [comment]
@rossabaker Thanks, I've noticed some strange behavior when wrapping the service.

I noticed the original issue after updating some dependencies and getting a `java.lang.NoSuchMethodError()`. Calling `handleWith` on the `Task` does not catch the error, it falls through to the default handler. Wrapping it in a `try/catch` calling `unsafePerformSync` and lifting it back in to a `Task` using `Task.now` does get the error. Strangely `Task.delay` instead of `Task.now` also fails. 

I'm not sure if you have any ideas? Looking through the code `Task.handleWith` should work and does work outside of HTTP4s so not sure whats happening.

```
import org.http4s._
import org.http4s.dsl._
import org.specs2.Specification
import org.specs2.matcher.TaskMatchers
import scalaz.concurrent.Task

class MainSpec extends Specification with TaskMatchers {
  def is = s2"""
  errorMiddleware should
    catch service errors using tryCatch        $tryCatch
    catch service errors using Task.handleWith $handleWith


  Task.handleWith should
    handle java.lang.NoSuchMethodError         $task
  """

  def haveBody(body: String) =
    returnValue(be_==(body)) ^^ { (resp: Task[Response]) => resp.flatMap(_.as[String]) }

  val service = HttpService.lift { request =>
    throw new java.lang.NoSuchMethodError()
  }

  // Fails, drops through to default handler when running in blaze, hands when test case enabled here
  def handleWith = {
    val errorMiddleWare = (service: HttpService) => HttpService.lift { req =>
      service(req).handleWith {
        case _: Throwable => InternalServerError("Caught error")
      }
    }

    errorMiddleWare(service)(Request()) must haveBody("Caught error")
  }

  // Works
  def tryCatch = {
    val errorMiddleWare = (service: HttpService) => HttpService.lift { req =>
      try {
        Task.now(service(req).unsafePerformSync)
      } catch {
        case _: Throwable => InternalServerError("Caught error")
      }
    }

    errorMiddleWare(service)(Request()) must haveBody("Caught error")
  }

  // Works
  def task = {
    Task.delay[Int](throw new java.lang.NoSuchMethodError()).handleWith {
      case _: Throwable => Task.now(1)
    } must returnValue(be_==(1))
  }
}
```

Currently using http4s version `0.15.13a`

**Comment by ChristopherDavenport** [comment]
I'll throw together a fix for this.

**Comment by rossabaker** [comment]
The `NoSuchMethodError` is not happening inside a `Task`, so the exception is thrown rather than wrapped in a failure.  `lift` could arguably start catching exceptions that happen outside tasks, but I'd consider any exceptions outside a Task to be a programming error, and I'd be very hesitant to swallow fatal errors like `NoSuchMethodError` into a task.  There's little we can do about these other than sadly render a 500, but there's no reason we couldn't have an uncaughtErrorHandler property on the server builders to customize the rendering and logging thereof.

**Comment by ChristopherDavenport** [comment]
Closed by #1314
